vistautils/parameters.py

```diff
@@ -1166,77 +1221,165 @@ def _check_all_keys_strings(mapping: Mapping, path=None):
+        check_arg(
+            not context.namespace_prefix, "Cannot interpolate with non-top-level context"
+        )
+
+        # These will be used when we represent parameters as dicts-of-dicts below.
+        def get_from_nested_dict(
+            nested_dict: Dict[str, Any], param_name: str
+        ) -> Optional[Any]:
+            parts = param_name.split(".")
+            cur_dict = nested_dict
+            for part in parts[:-1]:
+                if part in cur_dict:
+                    cur_dict = cur_dict[part]
+                else:
+                    return None
+            param_name = parts[-1]
+            return cur_dict.get(param_name, None)
+
+        def set_in_nested_dict(
+            nested_dict: Dict[str, Any], fully_qualified_param_name: str, value: Any
+        ) -> None:
+            parts = fully_qualified_param_name.split(".")
+            cur_dict = nested_dict
+            for part in parts[:-1]:
+                if part in cur_dict:
+                    cur_dict = cur_dict[part]
+                else:
+                    return None
+            param_name = parts[-1]
+            cur_dict[param_name] = value
+
+        # We make a mutable representation of the parameters we are interpolating
+        # as nested dictionaries in order to perform the actual interpolation.
+        # We will convert back to immutable Parameters objects at the end.
+        mutable_parameters = to_interpolate.as_nested_dicts()
+
         # In order to perform arbitrary interpolation, we perform topological sort on a graph where
         # the nodes are parameter keys, and edges point from those keys to each interpolation group
         # in that key's value. For example, the parameter entry `foo: %bar%/projects/%meep.baz%`
         # would give the edges (foo, bar) and (foo, meep.baz).
         #
         # pylint:disable=protected-access
-        nodes = list(to_interpolate._data.keys())
+        # Perform the interpolation in-place.
+        nodes = list()
         edges = list()
-        for key, val in to_interpolate._data.items():
-            if isinstance(val, str):
-                for interp_match in YAMLParametersLoader._INTERPOLATION_REGEX.findall(
-                    val
-                ):
-                    nodes.append(interp_match)
-                    edges.append((key, interp_match))
+
+        def gather_interpolation_edges(params: Parameters) -> None:
+            for key, val in params.namespaced_items():
+                if isinstance(val, str):
+                    for interp_match in YAMLParametersLoader._INTERPOLATION_REGEX.findall(
+                        val
+                    ):
+                        nodes.append(key)
+                        if get_from_nested_dict(mutable_parameters, interp_match):
+                            # We don't want to include nodes from the context in the interpolation
+                            # ordering since they should note themselves appear in the output
```

Comment:
```suggestion
                            # ordering since they should themselves appear in the output
```
and I don't understand what's meant by "output". Is it just that we aren't interpolating any values in the context, and we only use the context for interpolating in the given params?